src/gurt/misc.c

```diff
@@ -947,149 +951,478 @@ d_rank_range_list_free(d_rank_range_list_t *range_list)
+ * \param[in]		name	name of the environment variable.
+ * \return			0 on success, a negative value on error.
+ */
+int
+d_getenv_uint64_t(uint64_t *uint64_val, const char *name)
+{
+	char              *env;
+	char              *endptr;
+	unsigned long long val;
+	int                rc;
+
+	assert(name != NULL);
+	assert(uint64_val != NULL);
+
+	pthread_rwlock_rdlock(&d_env_lock);
+	env = getenv(name);
+	if (env == NULL) {
+		rc = -DER_NONEXIST;
+		goto out;
+	}
+
+	if (!dis_unsigned_str(env)) {
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	errno = 0;
+	val   = strtoull(env, &endptr, 0);
+	if (errno != 0 || endptr == env || *endptr != '\0') {
+		rc = -DER_INVAL;
+		goto out;
 	}
 
-	/* Now do scanf, check that the number was matched, and there are no extra unmatched
-	 * characters at the end.
-	 */
-	matched = sscanf(env_val, "%"PRId64"%n", &new_val, &count);
-	if (matched == 1 && env_len == count) {
-		*val = new_val;
-		D_DEBUG(DB_TRACE, "ENV '%s' set to %"PRId64"\n", env, *val);
-		return -DER_SUCCESS;
+#if UINT64_MAX != ULLONG_MAX
+	assert(sizeof(uint64_t) < sizeof(unsigned long long));
+	if (val > UINT64_MAX) {
+		rc = -DER_INVAL;
+		goto out;
 	}
+#endif
+	*uint64_val = (uint64_t)val;
+	rc          = -DER_SUCCESS;
+
+out:
+	pthread_rwlock_unlock(&d_env_lock);
+
+	return rc;
+}
+
+/**
+ * Thread safe wrapper of the libc putenv() function.
+ *
+ * \param[in]	name	name of the environment variable.
+ * \return		0 on success, a negative value on error.
+ */
+int
+d_putenv(char *name)
+{
+	int env_errno;
+	int rc;
+
+	pthread_rwlock_wrlock(&d_env_lock);
+	errno     = 0;
+	rc        = putenv(name);
+	env_errno = errno;
+	pthread_rwlock_unlock(&d_env_lock);
+
+	errno = env_errno;
+	return rc;
+}
+
+/**
+ * Thread safe wrapper of the libc setenv() function.
+ *
+ * \param[in]	name		name of the environment variable.
+ * \param[in]	value		value of the environment variable.
+ * \param[in]	overwrite	overwrite when nonzero.
+ * \return			0 on success, a negative value on error.
+ */
+int
+d_setenv(const char *name, const char *value, int overwrite)
+{
+	int env_errno;
+	int rc;
+
+	pthread_rwlock_wrlock(&d_env_lock);
+	errno     = 0;
+	rc        = setenv(name, value, overwrite);
+	env_errno = errno;
+	pthread_rwlock_unlock(&d_env_lock);
+
+	errno = env_errno;
+	return rc;
+}
+
+/**
+ * Thread safe wrapper of the libc unsetenv() function.
+ *
+ * \param[in]	name		name of the environment variable.
+ * \return			0 on success, a negative value on error.
+ */
+int
+d_unsetenv(const char *name)
+{
+	int env_errno;
+	int rc;
+
+	pthread_rwlock_wrlock(&d_env_lock);
+	errno     = 0;
+	rc        = unsetenv(name);
+	env_errno = errno;
+	pthread_rwlock_unlock(&d_env_lock);
+
+	errno = env_errno;
+	return rc;
+}
+
+/**
+ * Thread safe wrapper of the libc clearenv() function.
+ *
+ * \param[in]	name		name of the environment variable.
+ * \return			0 on success, a negative value on error.
+ */
+int
+d_clearenv(void)
+{
+	int rc;
+
+	pthread_rwlock_wrlock(&d_env_lock);
```

Comment:
Can you elaborate why this is safer than the hook version?   I think the issue we have seen is due to threads calling getenv in other libraries, which in this case would not be locked with respect to calls in daos libraries.  Whereas the hooked version would ensure the lock is acquired for all calls, no?

Reply:
The hook version can have some side effect with other third party library such as mvapich and eventually lead to segfault.  You could find in the following comment a detailed description of such situation https://daosio.atlassian.net/browse/DAOS-14532?focusedCommentId=121770

I tried to overcome this issue with updating hook or playing with library constructor.  At the end, I did not find an acceptable solution which could not have some side effect with other third party libraries.

After some discussion with @johannlombardi and @gnailzenh , it appeared that the least bad solution was to have our own thread safe functions for managing environment.  From my side, the only disadvantage of this solution is to be much more intrusive in our code.  Happily, we already have some dedicated non thread safe functions for managing the environment.

Reply:
we see an intermittent issue with fio that manifests as follows
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007d9faabc195d in getenv () from /lib64/libc.so.6
[Current thread is 1 (Thread 0x7d9fad4335c0 (LWP 60560))]
Missing separate debuginfos, use: yum debuginfo-install boost-atomic-1.66.0-13.el8.x86_64 boost-chrono-1.66.0-13.el8.x86_64 boost-date-time-1.66.0-13.el8.x86_64 boost-iostreams-1.66.0-13.el8.x86_64 boost-program-options-1.66.0-13.el8.x86_64 boost-random-1.66.0-13.el8.x86_64 boost-regex-1.66.0-13.el8.x86_64 boost-system-1.66.0-13.el8.x86_64 boost-thread-1.66.0-13.el8.x86_64 daxctl-libs-71.1-4.el8.x86_64 glibc-2.28-225.el8.x86_64 libaio-0.3.112-1.el8.x86_64 libblkid-2.32.1-42.el8_8.x86_64 libgcc-8.5.0-18.el8.x86_64 libibverbs-44.0-2.el8.1.x86_64 libicu-60.3-2.el8_1.x86_64 libmount-2.32.1-42.el8_8.x86_64 libnl3-3.7.0-1.el8.x86_64 libpmem-1.6.1-1.el8.x86_64 libpmemblk-1.6.1-1.el8.x86_64 librados2-12.2.7-9.el8.x86_64 librbd1-12.2.7-9.el8.x86_64 librdmacm-44.0-2.el8.1.x86_64 libselinux-2.9-8.el8.x86_64 libstdc++-8.5.0-18.el8.x86_64 libyaml-0.1.7-5.el8.x86_64 lz4-libs-1.8.3-3.el8_4.x86_64 ndctl-libs-71.1-4.el8.x86_64 nspr-4.34.0-3.el8_6.x86_64 nss-util-3.79.0-11.el8_7.x86_64 numactl-libs-2.0.12-13.el8.x86_64 openssl-libs-1.1.1k-9.el8_7.x86_64 pcre2-10.32-3.el8_6.x86_64 protobuf-c-1.3.0-6.el8.x86_64 systemd-libs-239-74.el8_8.x86_64 zlib-1.2.11-21.el8_7.x86_64
(gdb) bt
#0  0x00007d9faabc195d in getenv () from /lib64/libc.so.6
#1  0x00005b1f2183098c in check_status_file () at stat.c:2372
#2  0x00005b1f21839c25 in check_for_running_stats () at stat.c:2411
#3  0x00005b1f2186cf51 in do_usleep (usecs=10000) at backend.c:2460
#4  run_threads (sk_out=sk_out@entry=0x0) at backend.c:2460
#5  0x00005b1f2186d5b2 in fio_backend (sk_out=0x0) at backend.c:2513
#6  0x00005b1f2181c595 in main (argc=16, argv=0x7ffe8dcfb228, envp=<optimized out>) at fio.c:60

It doesn't show dlopen in the callstack.  Do you think it's related to this issue?

Reply:
There is a high probably that it could be related: however it is hard for me to be 100% sure without a more deep investigation in the core file and the related code source.
As for @daltonbohning issue with mvapich, you could use the full patch  #13250 to check if it also solves your issue with fio.